src/Interpreters/convertFieldToType.cpp

```diff
@@ -192,6 +192,16 @@ Field convertFieldToTypeImpl(const Field & src, const IDataType & type, const ID
     {
         return static_cast<const DataTypeDateTime &>(type).getTimeZone().fromDayNum(DayNum(src.get<Int32>()));
     }
+    else if (which_type.isDateTime64() && which_from_type.isDate())
+    {
+        const auto value = static_cast<const DataTypeDateTime64 &>(type).getTimeZone().fromDayNum(DayNum(src.get<UInt64>()));
+        return DecimalUtils::decimalFromComponentsWithMultiplier<DateTime64>(value, 0, 1);
```

Comment:
it looks like you need to use `DecimalBase::getScaleMultiplier()` instead of `1` here and below. Otherwise the produced returned value is always as if it was made for `DateTime64(1)`.